siobrultech_protocols/gem/protocol.py

```diff
@@ -225,6 +247,85 @@ def __str__(self) -> str:
         return f"Expected state to be {expected_str}; but got {self._actual.name}!"
 
 
+# Argument type of an ApiCall.
+T = TypeVar("T")
+# Return type of an ApiCall response parser.
+R = TypeVar("R")
+
+
+class ApiType(Enum):
+    ECM = 1
+    GEM = 2
+
+
+@dataclass
+class ApiCall(Generic[T, R]):
+    """
+    Helper class for making API calls with BidirectionalProtocol. There is one instance of
+    this class for each supported API call. This class handles the send_api_request and
+    receive_api_response parts of driving the protocol, since those are specific to each API
+    request type.
+    """
+
+    def __init__(
+        self,
+        gem_formatter: Callable[[T], str],
+        gem_parser: Callable[[str], R | None] | None,
+        ecm_formatter: Callable[[T], List[bytes]] | None,
+        ecm_parser: Callable[[bytes], R | None] | None,
+    ) -> None:
+        """
+        Create a new APICall.
+
+        gem_formatter - a callable that, given a parameter of type T, returns the string to send to the GEM to make the API call
+        gem_parser - a callable that, given a string, parses it into a value of type R.
+                    If there is not enough data to parse yet, it should return None.
+                    If there is enough data to parse, but it is malformed, it should raise an Exception.
+        ecm_formatter - a callable that, given a parameter of type T, returns the series of bytes chunks to send to the ECM to make the API call
+        ecm_parser - a callable that, given a bytes, parses it into a value of type R.
+                    If there is not enough data to parse yet, it should return None.
+                    If there is enough data to parse, but it is malformed, it should raise an Exception.
+        """
+        self._gem_formatter = gem_formatter
+        self._gem_parser = gem_parser
+        self._ecm_formatter = ecm_formatter
+        self._ecm_parser = ecm_parser
+
+    def format(
+        self,
+        api_type: ApiType,
+        arg: T,
+        serial_number: int | None,
+    ) -> List[bytes]:
+        if api_type == ApiType.GEM:
+            result = self._gem_formatter(arg)
+            if serial_number:
+                result = result.replace(
+                    ESCAPE_SEQUENCE,
+                    f"{TARGET_SERIAL_NUMBER_PREFIX}{serial_number%100000:05}",
+                )
+
+            return [result.encode()]
+        else:
```

Comment:
maybe check the type here and assert if we get something the code doesn't handle in an `else`?  While they might not ever release another device, catching this will help any future maintainer find their logic bug.